daliuge-engine/dlg/apps/dockerapp.py

```diff
@@ -401,12 +397,30 @@ def rm(container):
             stderr = container.logs(stream=False, stdout=False, stderr=True)
             msg = "Container %s didn't finish successfully (exit code %d)" % (cId, self._exitCode)
             logger.error(msg + ", output follows.\n==STDOUT==\n%s==STDERR==\n%s", stdout, stderr)
-            rm(container)
             raise Exception(msg)
 
-        rm(container)
         c.api.close()
 
+    @overrides
+    def setCompleted(self):
+        super(BarrierAppDROP, self).setCompleted()
+        self.container.stop()
+
+    @overrides
+    def setError(self):
+        super(BarrierAppDROP, self).setError()
+        self.container.stop()
+
+    @overrides
+    def cancel(self):
+        super(BarrierAppDROP, self).cancel()
+        self.container.stop()
```

Comment:
The idea here is cancelling a session (or basically changing its state after running) will call `container.stop()` whilst another co-routine is blocked by `container.wait()`